Initial implementation of check-upgrade

evan-goode opened the pull request:
So far just a start, missing --changelog, security filters, obsoletes.

Uses @m-blaha's `package_list_sections.cpp` from #298 for now, it would be nice to reuse it, but `check-upgrade` might want its own logic for printing its output.

evan-goode commented:
`--changelogs`, advisory filters, and the obsoletes table are now implemented. Tests still on the way, but this is ready for review.

evan-goode commented on libdnf/repo/solv_repo.cpp:
```diff
@@ -421,7 +421,8 @@ void SolvRepo::load_system_repo(const std::string & rootdir) {
 
-    int flagsrpm = REPO_REUSE_REPODATA | RPM_ADD_WITH_HDRID | REPO_USE_ROOTDIR;
+    // TODO(egoode) investigate performance hit of RPM_ADD_WITH_CHANGELOG, possibly make this configurable
+    int flagsrpm = REPO_REUSE_REPODATA | RPM_ADD_WITH_HDRID | REPO_USE_ROOTDIR | RPM_ADD_WITH_CHANGELOG;
```
This change causes changelog data to always be loaded for the system repository. I'm not sure what the performance implications are; there is no measurable difference on my system, but I don't have a lot of packages installed. We may want to go ahead and make this function more configurable, or we can go ahead with this change and possibly revisit in the future.

jan-kolarik replied:
I guess the TODO is OK there and when we'll encounter any related issue in the future, it could be revisited then as you've written. Another related issue is that we load all metadata for the system repository when calling `Repo::load_system_repo()`, and there is already a TODO present regarding this.

evan-goode commented:
Resolves https://github.com/rpm-software-management/dnf5/issues/134

jan-kolarik commented on libdnf/conf/config.h:
```diff
@@ -0,0 +1,26 @@
+/*
```
This file should be dropped as it is automatically generated.

jan-kolarik replied:
Hmm, I guess so. I didn't notice that before as I have this file ignored in my personal `.gitignore_global`, probably already since started to work with the `libdnf` repo where it's also not in the ignore file. But it should be, I don't see a reason why not...

jan-kolarik commented on include/libdnf-cli/exception.hpp:
```diff
@@ -88,6 +88,29 @@ class CommandExitError : public Error {
     int exit_code;
 };
 
+/// Exception used when non-standard exit code should be returned
```
Maybe just improve the doc string as it is the same as with `CommandExitError` - add an explanation what the `Silent` prefix means...

evan-goode replied:
Ah, meant to change this, thank you.

jan-kolarik commented on libdnf-cli/output/changelogs.cpp:
```diff
@@ -0,0 +1,118 @@
+
+namespace libdnf::cli::output {
+
+void print_changelogs(
```
There are more things that IMHO could be refactored, but I understand this was only moved from different file.

jan-kolarik commented on dnf5/commands/check-upgrade/check-upgrade.cpp:
```diff
@@ -0,0 +1,191 @@
+    cmd.set_description("Check for available package upgrades");
+
+    changelogs = dynamic_cast<libdnf::OptionBool *>(
+        parser.add_init_value(std::unique_ptr<libdnf::OptionBool>(new libdnf::OptionBool(false))));
```
If think this could be done in shorter way like this: `std::make_unique<libdnf::OptionBool>(false)`.

evan-goode replied:
Sure, thanks!

jan-kolarik commented on dnf5/commands/check-upgrade/check-upgrade.cpp:
```diff
@@ -0,0 +1,191 @@
+    auto & cmd = *get_argument_parser_command();
+    cmd.set_description("Check for available package upgrades");
+
+    changelogs = dynamic_cast<libdnf::OptionBool *>(
```
Probably nothing serious, but is really the `dynamic_cast` necessary if we would separate the `unique_ptr` creation and raw pointer assignments?

jan-kolarik replied:
I was thinking about not using the `add_init_value` return value at all, because it only returns the argument that was passed in. The overall usage of that function seems kinda weird to me, because it seems only moving the ownership of options to the parser and nothing else. And just because of that there are then additional runtime casting etc... But anyway, you're right that it's used like that on many more places, so it would be better to do some systemic change about that.

jan-kolarik commented on dnf5/commands/check-upgrade/check-upgrade.cpp:
```diff
@@ -0,0 +1,191 @@
+    auto & ctx = get_context();
+    auto & config = ctx.base.get_config();
+
+    libdnf::rpm::PackageQuery full_package_query(ctx.base);
```
This is not used anywhere in this function. Could be probably created later inside the `print_changelogs`? So then this parameter would be dropped.

jan-kolarik commented on dnf5/commands/check-upgrade/check-upgrade.cpp:
```diff
@@ -0,0 +1,191 @@
+    auto & ctx = get_context();
+    auto & config = ctx.base.get_config();
+
+    libdnf::rpm::PackageQuery full_package_query(ctx.base);
```
As it is not used anywhere before the passing to the `print_changelogs`, I would create it right before that. I think it's better to introduce the object at the time it is actually used/needed.

evan-goode replied:
Thanks, I'll move this.

evan-goode commented:
> Comparing the output of `dnf check-upgrade` and `dnf5 check-upgrade`, there seem to be all available upgrades to a given package in the list, not just the latest one, f.e.:
> 
> ```
> createrepo_c.x86_64                                0.20.1-20230329010146.26.g97242bc.fc36  copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
> createrepo_c.x86_64                                0.20.1-20230330005723.26.g97242bc.fc36  copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
> createrepo_c.x86_64                                0.20.1-20230331005638.26.g97242bc.fc36  copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
> createrepo_c.x86_64                                0.20.1-20230401005540.26.g97242bc.fc36  copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
> createrepo_c.x86_64                                0.20.1-20230402005729.26.g97242bc.fc36  copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
> createrepo_c.x86_64                                0.20.1-20230403005322.26.g97242bc.fc36  copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
> createrepo_c.x86_64                                0.20.1-20230404005315.26.g97242bc.fc36  copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
> createrepo_c.x86_64                                0.21.1-20230405004456.0.g0652d73.fc36   copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
> createrepo_c.x86_64                                0.21.1-20230406005151.0.g0652d73.fc36   copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
> createrepo_c.x86_64                                0.21.1-20230407004839.0.g0652d73.fc36   copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
> createrepo_c.x86_64                                0.21.1-20230408005107.0.g0652d73.fc36   copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
> createrepo_c.x86_64                                0.21.1-20230409005451.0.g0652d73.fc36   copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
> createrepo_c.x86_64                                0.21.1-20230410005321.0.g0652d73.fc36   copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
> createrepo_c.x86_64                                0.21.1-20230411005345.0.g0652d73.fc36   copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
> ```

Thanks, I didn't see this testing with just the standard repos. Fixed now.

Sorry for keeping adding more commits, when I fixed one test it happened to break another and it took me a second to notice.

jan-kolarik commented:
> > Comparing the output of `dnf check-upgrade` and `dnf5 check-upgrade`, there seem to be all available upgrades to a given package in the list, not just the latest one, f.e.:
> > ```
> > createrepo_c.x86_64                                0.20.1-20230329010146.26.g97242bc.fc36  copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
> > createrepo_c.x86_64                                0.20.1-20230330005723.26.g97242bc.fc36  copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
> > createrepo_c.x86_64                                0.20.1-20230331005638.26.g97242bc.fc36  copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
> > createrepo_c.x86_64                                0.20.1-20230401005540.26.g97242bc.fc36  copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
> > createrepo_c.x86_64                                0.20.1-20230402005729.26.g97242bc.fc36  copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
> > createrepo_c.x86_64                                0.20.1-20230403005322.26.g97242bc.fc36  copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
> > createrepo_c.x86_64                                0.20.1-20230404005315.26.g97242bc.fc36  copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
> > createrepo_c.x86_64                                0.21.1-20230405004456.0.g0652d73.fc36   copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
> > createrepo_c.x86_64                                0.21.1-20230406005151.0.g0652d73.fc36   copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
> > createrepo_c.x86_64                                0.21.1-20230407004839.0.g0652d73.fc36   copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
> > createrepo_c.x86_64                                0.21.1-20230408005107.0.g0652d73.fc36   copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
> > createrepo_c.x86_64                                0.21.1-20230409005451.0.g0652d73.fc36   copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
> > createrepo_c.x86_64                                0.21.1-20230410005321.0.g0652d73.fc36   copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
> > createrepo_c.x86_64                                0.21.1-20230411005345.0.g0652d73.fc36   copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
> > ```
> 
> Thanks, I didn't see this testing with just the standard repos. Fixed now.
> 
> Sorry for keeping adding more commits, when I fixed one test it happened to break another and it took me a second to notice.

Hmm, I am still getting this output even with the fresh changes ...

jan-kolarik commented:
Now the list of upgrades looks correct and same with DNF and DNF5, but the obsoleting packages are very different.

DNF:
```
Obsoleting Packages
kernel-headers.x86_64                                                         6.2.6-100.fc36                                                     updates                                                          
    kernel-headers.x86_64                                                     5.18.13-200.fc36                                                   @updates                                                         
python3-dnf-plugins-core.noarch                                               4.3.1-20230404005320.17.g48b29df.fc36                              copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly 
    python3-dnf-plugin-system-upgrade.noarch                                  4.0.16-20220815011335.0.gcef43db.fc36                              @copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
python3-dnf-plugins-core.noarch                                               4.3.1-20230404005320.17.g48b29df.fc36                              copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly 
    python3-dnf-plugins-core.noarch                                           4.2.1-1.fc36                                                       @updates                                                         
python3-dnf-plugins-core.noarch                                               4.3.1-20230405004502.17.g48b29df.fc36                              copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly 
    python3-dnf-plugin-system-upgrade.noarch                                  4.0.16-20220815011335.0.gcef43db.fc36                              @copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
python3-dnf-plugins-core.noarch                                               4.3.1-20230405004502.17.g48b29df.fc36                              copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly 
    python3-dnf-plugins-core.noarch                                           4.2.1-1.fc36                                                       @updates                                                         
python3-dnf-plugins-core.noarch                                               4.4.0-20230406005156.0.g35f9595.fc36                               copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly 
    python3-dnf-plugin-system-upgrade.noarch                                  4.0.16-20220815011335.0.gcef43db.fc36                              @copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
python3-dnf-plugins-core.noarch                                               4.4.0-20230406005156.0.g35f9595.fc36                               copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly 
    python3-dnf-plugins-core.noarch                                           4.2.1-1.fc36                                                       @updates                                                         
python3-dnf-plugins-core.noarch                                               4.4.0-20230407004844.0.g35f9595.fc36                               copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly 
    python3-dnf-plugin-system-upgrade.noarch                                  4.0.16-20220815011335.0.gcef43db.fc36                              @copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
python3-dnf-plugins-core.noarch                                               4.4.0-20230407004844.0.g35f9595.fc36                               copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly 
    python3-dnf-plugins-core.noarch                                           4.2.1-1.fc36                                                       @updates                                                         
python3-dnf-plugins-core.noarch                                               4.4.0-20230408005113.0.g35f9595.fc36                               copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly 
    python3-dnf-plugin-system-upgrade.noarch                                  4.0.16-20220815011335.0.gcef43db.fc36                              @copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
python3-dnf-plugins-core.noarch                                               4.4.0-20230408005113.0.g35f9595.fc36                               copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly 
    python3-dnf-plugins-core.noarch                                           4.2.1-1.fc36                                                       @updates                                                         
python3-dnf-plugins-core.noarch                                               4.4.0-20230409005456.0.g35f9595.fc36                               copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly 
    python3-dnf-plugin-system-upgrade.noarch                                  4.0.16-20220815011335.0.gcef43db.fc36                              @copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
python3-dnf-plugins-core.noarch                                               4.4.0-20230409005456.0.g35f9595.fc36                               copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly 
    python3-dnf-plugins-core.noarch                                           4.2.1-1.fc36                                                       @updates                                                         
python3-dnf-plugins-core.noarch                                               4.4.0-20230410005326.0.g35f9595.fc36                               copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly 
    python3-dnf-plugin-system-upgrade.noarch                                  4.0.16-20220815011335.0.gcef43db.fc36                              @copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
python3-dnf-plugins-core.noarch                                               4.4.0-20230410005326.0.g35f9595.fc36                               copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly 
    python3-dnf-plugins-core.noarch                                           4.2.1-1.fc36                                                       @updates                                                         
python3-dnf-plugins-core.noarch                                               4.4.0-20230411005351.0.g35f9595.fc36                               copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly 
    python3-dnf-plugin-system-upgrade.noarch                                  4.0.16-20220815011335.0.gcef43db.fc36                              @copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
python3-dnf-plugins-core.noarch                                               4.4.0-20230411005351.0.g35f9595.fc36                               copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly 
    python3-dnf-plugins-core.noarch                                           4.2.1-1.fc36                                                       @updates                                                         
python3-dnf-plugins-core.noarch                                               4.4.0-20230412005337.0.g35f9595.fc36                               copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly 
    python3-dnf-plugin-system-upgrade.noarch                                  4.0.16-20220815011335.0.gcef43db.fc36                              @copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
python3-dnf-plugins-core.noarch                                               4.4.0-20230412005337.0.g35f9595.fc36                               copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly 
    python3-dnf-plugins-core.noarch                                           4.2.1-1.fc36                                                       @updates                                                         
python3-dnf-plugins-core.noarch                                               4.4.0-20230413005210.0.g35f9595.fc36                               copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly 
    python3-dnf-plugin-system-upgrade.noarch                                  4.0.16-20220815011335.0.gcef43db.fc36                              @copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
python3-dnf-plugins-core.noarch                                               4.4.0-20230413005210.0.g35f9595.fc36                               copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly 
    python3-dnf-plugins-core.noarch                                           4.2.1-1.fc36                                                       @updates                                                         
python3-dnf-plugins-core.noarch                                               4.4.0-20230415005515.0.g35f9595.fc36                               copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly 
    python3-dnf-plugin-system-upgrade.noarch                                  4.0.16-20220815011335.0.gcef43db.fc36                              @copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
python3-dnf-plugins-core.noarch                                               4.4.0-20230415005515.0.g35f9595.fc36                               copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly 
    python3-dnf-plugins-core.noarch                                           4.2.1-1.fc36                                                       @updates                                                         
python3-dnf-plugins-core.noarch                                               4.4.0-20230416005543.0.g35f9595.fc36                               copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly 
    python3-dnf-plugin-system-upgrade.noarch                                  4.0.16-20220815011335.0.gcef43db.fc36                              @copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
python3-dnf-plugins-core.noarch                                               4.4.0-20230416005543.0.g35f9595.fc36                               copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly 
    python3-dnf-plugins-core.noarch                                           4.2.1-1.fc36                                                       @updates                                                         
python3-dnf-plugins-core.noarch                                               4.4.0-20230417005447.0.g35f9595.fc36                               copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly 
    python3-dnf-plugin-system-upgrade.noarch                                  4.0.16-20220815011335.0.gcef43db.fc36                              @copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
python3-dnf-plugins-core.noarch                                               4.4.0-20230417005447.0.g35f9595.fc36                               copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly 
    python3-dnf-plugins-core.noarch                                           4.2.1-1.fc36                                                       @updates                                                         
python3-dnf-plugins-extras-common.noarch                                      4.0.17-1.fc36                                                      updates                                                          
    python3-dnf-plugins-extras-common.noarch                                  4.0.16-20220815011335.0.gcef43db.fc36                              @copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
python3-dnf-plugins-extras-common.noarch                                      4.0.17-20230404005321.3.g2c54189.fc36                              copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly 
    python3-dnf-plugins-extras-common.noarch                                  4.0.16-20220815011335.0.gcef43db.fc36                              @copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
python3-dnf-plugins-extras-common.noarch                                      4.0.17-20230405004503.3.g2c54189.fc36                              copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly 
    python3-dnf-plugins-extras-common.noarch                                  4.0.16-20220815011335.0.gcef43db.fc36                              @copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
python3-dnf-plugins-extras-common.noarch                                      4.1.0-20230406005157.0.gaee8265.fc36                               copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly 
    python3-dnf-plugins-extras-common.noarch                                  4.0.16-20220815011335.0.gcef43db.fc36                              @copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
python3-dnf-plugins-extras-common.noarch                                      4.1.0-20230407004845.0.gaee8265.fc36                               copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly 
    python3-dnf-plugins-extras-common.noarch                                  4.0.16-20220815011335.0.gcef43db.fc36                              @copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
python3-dnf-plugins-extras-common.noarch                                      4.1.0-20230408005113.0.gaee8265.fc36                               copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly 
    python3-dnf-plugins-extras-common.noarch                                  4.0.16-20220815011335.0.gcef43db.fc36                              @copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
python3-dnf-plugins-extras-common.noarch                                      4.1.0-20230409005456.0.gaee8265.fc36                               copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly 
    python3-dnf-plugins-extras-common.noarch                                  4.0.16-20220815011335.0.gcef43db.fc36                              @copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
python3-dnf-plugins-extras-common.noarch                                      4.1.0-20230410005326.0.gaee8265.fc36                               copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly 
    python3-dnf-plugins-extras-common.noarch                                  4.0.16-20220815011335.0.gcef43db.fc36                              @copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
python3-dnf-plugins-extras-common.noarch                                      4.1.0-20230411005352.0.gaee8265.fc36                               copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly 
    python3-dnf-plugins-extras-common.noarch                                  4.0.16-20220815011335.0.gcef43db.fc36                              @copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
python3-dnf-plugins-extras-common.noarch                                      4.1.0-20230412005338.0.gaee8265.fc36                               copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly 
    python3-dnf-plugins-extras-common.noarch                                  4.0.16-20220815011335.0.gcef43db.fc36                              @copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
python3-dnf-plugins-extras-common.noarch                                      4.1.0-20230413005210.0.gaee8265.fc36                               copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly 
    python3-dnf-plugins-extras-common.noarch                                  4.0.16-20220815011335.0.gcef43db.fc36                              @copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
python3-dnf-plugins-extras-common.noarch                                      4.1.0-20230415005516.0.gaee8265.fc36                               copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly 
    python3-dnf-plugins-extras-common.noarch                                  4.0.16-20220815011335.0.gcef43db.fc36                              @copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
python3-dnf-plugins-extras-common.noarch                                      4.1.0-20230416005543.0.gaee8265.fc36                               copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly 
    python3-dnf-plugins-extras-common.noarch                                  4.0.16-20220815011335.0.gcef43db.fc36                              @copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
python3-dnf-plugins-extras-common.noarch                                      4.1.0-20230417005447.0.gaee8265.fc36                               copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly 
    python3-dnf-plugins-extras-common.noarch                                  4.0.16-20220815011335.0.gcef43db.fc36                              @copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
```
DNF5:
```
Obsoleting packages
kernel-headers.x86_64                              6.2.6-100.fc36                         updates
    kernel-headers.x86_64                          5.18.13-200.fc36                       updates
python3-dnf-plugins-core.noarch                    4.4.0-20230418005241.0.g35f9595.fc36   copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
    python3-dnf-plugin-system-upgrade.noarch       4.0.16-20220815011335.0.gcef43db.fc36  copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
    python3-dnf-plugins-core.noarch                4.2.1-1.fc36                           updates
python3-dnf-plugins-extras-common.noarch           4.1.0-20230418005241.0.gaee8265.fc36   copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
    python3-dnf-plugins-extras-common.noarch       4.0.16-20220815011335.0.gcef43db.fc36  copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf-nightly
```

But I think the DNF5 implemention behaves correct and consistent in this matter as we are showing the latest available update for each package and the obsoleting list is also reflecting this behavior. Compared to the DNF where the obsoleting list shows all available package updates... What do you think?

evan-goode commented:
Right, I noticed that also. I definitely prefer the DNF5 output. It doesn't make sense to list all of the obsoleted versions, especially for repos such as our COPR that have many available versions.
